internal/ledgerutil/compare.go

```diff
@@ -69,32 +72,81 @@ func Compare(snapshotDir1 string, snapshotDir2 string, outputDirLoc string, firs
 		return 0, "", errors.Errorf("%s already exists in %s. Choose a different location or remove the existing results. Aborting compare", outputDirName, outputDirLoc)
 	}
 
-	// Create the output files
-	allOutputFile, err := newJSONFileWriter(filepath.Join(outputDirPath, AllDiffsByKey))
-	if err != nil {
-		return 0, "", err
+	// Generate all public data differences between snapshots
+	if !equalPub {
+		snapshotPubReader1, err := privacyenabledstate.NewSnapshotReader(snapshotDir1,
+			privacyenabledstate.PubStateDataFileName, privacyenabledstate.PubStateMetadataFileName)
+		if err != nil {
+			return 0, "", err
+		}
+		snapshotPubReader2, err := privacyenabledstate.NewSnapshotReader(snapshotDir2,
+			privacyenabledstate.PubStateDataFileName, privacyenabledstate.PubStateMetadataFileName)
+		if err != nil {
+			return 0, "", err
+		}
+		outputPubFile, err := findDifferences(outputDirPath, AllPubDiffsByKey, snapshotPubReader1, snapshotPubReader2, firstDiffs)
+		if err != nil {
+			return 0, "", err
+		}
+		count += outputPubFile.count
 	}
 
-	// Create snapshot readers to read both snapshots
-	snapshotReader1, err := privacyenabledstate.NewSnapshotReader(snapshotDir1,
-		privacyenabledstate.PubStateDataFileName, privacyenabledstate.PubStateMetadataFileName)
-	if err != nil {
-		return 0, "", err
+	// Generate all private data differences between snapshots
+	if !equalPvt {
+		snapshotPvtReader1, err := privacyenabledstate.NewSnapshotReader(snapshotDir1,
+			privacyenabledstate.PvtStateHashesFileName, privacyenabledstate.PvtStateHashesMetadataFileName)
+		if err != nil {
+			return 0, "", err
+		}
+		snapshotPvtReader2, err := privacyenabledstate.NewSnapshotReader(snapshotDir2,
+			privacyenabledstate.PvtStateHashesFileName, privacyenabledstate.PvtStateHashesMetadataFileName)
+		if err != nil {
+			return 0, "", err
+		}
+		outputPvtFile, err := findDifferences(outputDirPath, AllPvtDiffsByKey, snapshotPvtReader1, snapshotPvtReader2, firstDiffs)
+		if err != nil {
+			return 0, "", err
+		}
+		count += outputPvtFile.count
 	}
-	snapshotReader2, err := privacyenabledstate.NewSnapshotReader(snapshotDir2,
-		privacyenabledstate.PubStateDataFileName, privacyenabledstate.PubStateMetadataFileName)
+
+	// Generate early differences output file
+	if firstDiffs != 0 {
+		firstDiffsOutputFile, err := newJSONFileWriter(filepath.Join(outputDirPath, FirstDiffsByHeight))
+		if err != nil {
+			return 0, "", err
+		}
+		sort.Sort(allRecords)
```

Comment:
This was initially the algorithm I had in mind but went with the current implementation because I thought time complexity would be more important. In hindsight, from the perspective that there could potentially be millions of found differences and wanting to minimize space complexity, this algorithm makes a lot more sense.